edb/server/ha/stolon.py

```diff
@@ -0,0 +1,243 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2021-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import asyncio
+import base64
+import functools
+import json
+import logging
+import ssl
+import urllib.parse
+from typing import *
+
+import httptools
+
+from . import base
+
+logger = logging.getLogger("edb.server")
+
+
+class StolonBackend(base.HABackend):
+    _master_addr: Optional[Tuple[str, int]]
+
+    def __init__(self):
+        self._waiter = None
+        self._master_addr = None
+        self._protocol = None
+        self._watching = False
+        self._cluster_protocol = None
+
+    async def get_cluster_consensus(self) -> Tuple[str, int]:
+        if self._master_addr is None:
+            started_by_us = await self.start_watching()
+            try:
+                assert self._waiter is None
+                loop = asyncio.get_running_loop()
+                self._waiter = loop.create_future()
+                await self._waiter
+            finally:
+                if started_by_us:
+                    self.stop_watching()
+        assert self._master_addr
+        return self._master_addr
+
+    async def start_watching(self, cluster_protocol=None):
+        if cluster_protocol is not None:
+            assert self._cluster_protocol is None
+            self._cluster_protocol = cluster_protocol
+        if not self._watching:
+            self._watching = True
+            try:
+                self._protocol = await self._start_watching()
+                return True
+            except BaseException:
+                self._watching = False
+                raise
+        return False
+
+    def stop_watching(self):
+        self._watching = False
+        self._cluster_protocol = None
+        protocol, self._protocol = self._protocol, None
+        if protocol is not None:
+            protocol.close()
+
+    def get_master_addr(self):
+        return self._master_addr
+
+    async def _start_watching(self):
+        raise NotImplementedError
+
+    def on_cluster_data(self, data):
+        cluster_status = data.get("cluster", {}).get("status", {})
+        master_db = cluster_status.get("master")
+        cluster_phase = cluster_status.get("phase")
+        if cluster_phase != "normal":
+            logger.debug("Stolon cluster phase: %s", cluster_phase)
```

Comment:
nit:
```suggestion
            logger.debug("Stolon cluster phase: %r", cluster_phase)
```

an so in other places; that will add quotes to strings making it easier to read logs